R/mcmc.R

```diff
@@ -1,280 +1,151 @@
-                                          n.iter=
-                                              (options@iterations - options@burnin),
-                                          thin=options@step,
-                                          progress.bar=
-                                              ifelse(verbose,
-                                                     'text',
-                                                     'none'))"
-
-                  ## evaluate with or without outstream capturing
-                  if(verbose)
-                  {
-                      eval(parse(text=samplesCode))
-                  } else {
-
-                      ## this is necessary because some outputs
-                      ## are written directly from the JAGS compiled
-                      ## code to the outstream
-                      capture.output(eval(parse(text=samplesCode)))
-                  }
+setMethod(
+  f = "mcmc",
+  signature = signature(
+    data = "GeneralData",
+    model = "GeneralModel",
+    options = "McmcOptions"
+  ),
+  def = function(data,
+                 model,
+                 options,
+                 verbose = FALSE,
+                 from_prior = data@nObs == 0L,
+                 ...) {
+    assert_flag(verbose)
+    assert_flag(from_prior)
+
+    if (verbose) {
+      futile.logger::flog.threshold(futile.logger::INFO, name = "mcmc")
+    } else {
+      futile.logger::flog.threshold(futile.logger::FATAL, name = "mcmc")
+    }
 
-                  ## reformat slightly for Samples object
-                  ret <- lapply(samples,
-                                function(x) {
-                                    ## take the first chain (because we use only
-                                    ## one anyway), and take all samples (burnin
-                                    ## was already not saved!)
-                                    x <- x[, , 1L]
-                                    ## transpose if it is a matrix
-                                    ## (in case that there are multiple parameters
-                                    ## in a node)
-                                    if(is.matrix(x))
-                                    {
-                                        x <- t(x)
-                                    }
-                                    x
-                                })
+    jags_model_fun <- if (from_prior) {
+      model@priormodel
+    } else {
+      h_join_models(model@datamodel, model@priormodel)
+    }
 
-              }
+    # Write the model into the file.
+    jags_dir <- file.path(tempdir(), "R_crmPack")
+    jags_file <- tempfile("jags_model_fun", jags_dir, ".txt")
+    # Don't warn, as the temp dir often exists (which is OK).
+    dir.create(jags_dir, showWarnings = FALSE)
+    futile.logger::flog.info("Temporary directory: %s", jags_dir, name = "mcmc")
+    h_write_model(jags_model_fun, jags_file)
```

Comment:
looking at this I think it might be better to move this stuff into `h_write_model` so that we can just write here 
```suggestion  
    jags_file <- h_write_model(jags_model_fun)
```